File: solutions/java/bottle-song/1/src/main/java/BottleSong.java
class BottleSong {
    static final String TEMPLATE = """
    %1$s green bottles hanging on the wall,
    %1$s green bottles hanging on the wall,
    And if one green bottle should accidentally fall,
    There'll be %2$s green bottles hanging on the wall.
    
    """;
    static final String[] NUMBERS = {"no", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten"};
    String recite(int startBottles, int takeDown) {
        StringBuilder sb = new StringBuilder();
        for (int i = startBottles; i > startBottles - takeDown; i--) {
            sb.append(String.format(TEMPLATE, NUMBERS[i], NUMBERS[i - 1].toLowerCase()).replaceAll("(?<=[oO]ne green bottle)s", ""));
        }
        sb.deleteCharAt(sb.length() - 1);
        return sb.toString();
    }

}

File: solutions/kotlin/bottle-song/1/src/main/kotlin/BottleSong.kt
// Dedicated to Junko F. Didi and Shree DR.MDD
object BottleSong {
    fun recite(startBottles : Int, takeDown : Int) : String {
        val lyrics = StringBuilder()
        for (cur in startBottles downTo startBottles - takeDown + 1){
            lyrics.append(stanza(cur))
            if (takeDown > 1 && cur - 1 != (startBottles - takeDown)){
                lyrics.append("\n")
            }
        }
        return lyrics.toString().trim()
    }

    fun stanza(count: Int): String{
        val words = arrayListOf("no", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten")
        val lineA = "%s green bottle" + (if (count != 1) "s" else "") + " hanging on the wall,\n"
        val lineB = "And if one green bottle should accidentally fall,\n"
        val lineC = "There'll be %s green bottle" + (if (count - 1 != 1) "s" else "") + " hanging on the wall.\n"
        return String.format(lineA, words[count]).repeat(2) + lineB + String.format(lineC, words[count - 1].lowercase())
    }
}
